File: szavazas-core/src/main/java/hu/kdea/szavazas/ballotprocessor/projection/EdgeReconstructor.java
package hu.kdea.szavazas.ballotprocessor.projection;

import hu.kdea.szavazas.ballotprocessor.GridConstants;
import java.util.ArrayList;
import java.util.List;
import kotlin.Pair;

public final class EdgeReconstructor {
    private EdgeReconstructor() {
    }

    public static List<Pair<Integer, Integer>> reconstructViaPairs(
        float[] projection,
        int offset,
        int expectedPairs,
        boolean emptySecondColumn
    ) {
        List<Integer> rawPeaks = PeakFinder.findRawPeaks(projection, offset);
        List<Integer> merged = PeakFinder.mergeClosePeaks(rawPeaks);
        int totalColumns = emptySecondColumn ? expectedPairs + 1 : expectedPairs;
        int span = projection.length;
        double averageColumnWidth = (double) span / totalColumns;
        int minGap = (int) (averageColumnWidth * 0.25);
        int maxGap = (int) (averageColumnWidth * 0.75);
        List<Pair<Integer, Integer>> pairs = pairEdges(merged, minGap, maxGap, 10);
        if (pairs.size() != expectedPairs) {
            return null;
        }
        for (int i = 0; i < pairs.size() - 1; i++) {
            if (pairs.get(i).getSecond() >= pairs.get(i + 1).getFirst()) {
                return null;
            }
        }
        if (emptySecondColumn && pairs.size() >= 2) {
            List<Integer> gaps = new ArrayList<>();
            for (int i = 0; i < pairs.size() - 1; i++) {
                gaps.add(pairs.get(i + 1).getFirst() - pairs.get(i).getFirst());
            }
            int firstGap = gaps.get(0);
            List<Integer> otherGaps = gaps.subList(1, gaps.size());
            if (otherGaps.isEmpty()) {
                return null;
            }
            double avgOtherGap = otherGaps.stream().mapToInt(Integer::intValue).average().orElse(0.0);
            if (Math.abs(firstGap - 2 * avgOtherGap) > avgOtherGap * 0.5) {
                return null;
            }
        }
        return pairs;
    }

    public static List<Pair<Integer, Integer>> getRawPairs(float[] projection, int offset) {
        List<Integer> rawPeaks = PeakFinder.findRawPeaks(projection, offset);
        List<Integer> merged = PeakFinder.mergeClosePeaks(rawPeaks);
        return pairEdges(merged, GridConstants.EXPECTED_WIDTH_MIN, GridConstants.EXPECTED_WIDTH_MAX, 10);
    }

    public static List<Pair<Integer, Integer>> pairEdges(List<Integer> peaks, int minGap, int maxGap) {
        return pairEdges(peaks, minGap, maxGap, 10);
    }

    public static List<Pair<Integer, Integer>> pairEdges(List<Integer> peaks, int minGap, int maxGap, int maxLookAhead) {
        List<Pair<Integer, Integer>> pairs = new ArrayList<>();
        boolean[] used = new boolean[peaks.size()];
        for (int i = 0; i < peaks.size(); i++) {
            if (used[i]) {
                continue;
            }
            for (int j = i + 1; j < Math.min(i + maxLookAhead, peaks.size()); j++) {
                int gap = peaks.get(j) - peaks.get(i);
                if (gap >= minGap && gap <= maxGap && !used[j]) {
                    pairs.add(new Pair<>(peaks.get(i), peaks.get(j)));
                    used[i] = true;
                    used[j] = true;
                    break;
                }
            }
        }
        return pairs;
    }
}

File: test-utils/src/main/java/hu/kdea/szavazas/ballotprocessor/projection/EdgeReconstructor.kt
package hu.kdea.szavazas.ballotprocessor

object EdgeReconstructor {

    /**
     * Reconstructs edges using the peak‑pairing algorithm.
     * Returns null if the number of pairs doesn't match expectedPairs,
     * or if the pairs are overlapping / not strictly increasing.
     */
    fun reconstructViaPairs(
        proj: FloatArray,
        offset: Int,
        expectedPairs: Int,
        emptySecondColumn: Boolean
    ): List<Pair<Int, Int>>? {
        val rawPeaks = PeakFinder.findRawPeaks(proj, offset)
        val merged = PeakFinder.mergeClosePeaks(rawPeaks)

        // Dynamic pair width: if emptySecondColumn, there is one extra (empty) column,
        // so the total number of columns = expectedPairs + 1.
        val totalColumns = if (emptySecondColumn) expectedPairs + 1 else expectedPairs
        val span = proj.size
        val avgColumnWidth = span.toDouble() / totalColumns
        // A pair (left‑right) typically occupies about one column width.
        // Use a reasonable fraction to allow some variation.
        val minGap = (avgColumnWidth * 0.25).toInt()
        val maxGap = (avgColumnWidth * 0.75).toInt()

        val pairs = pairEdges(merged, minGap, maxGap)

        if (pairs.size != expectedPairs) return null

        // Validate non‑overlapping, strictly increasing order
        for (i in 0 until pairs.size - 1) {
            if (pairs[i].second >= pairs[i + 1].first) return null
        }

        // For columns with an empty second slot, the gap between the first two pairs
        // must be roughly twice the average gap between the remaining pairs.
        if (emptySecondColumn && pairs.size >= 2) {
            val gaps = pairs.zipWithNext { a, b -> b.first - a.first }
            val firstGap = gaps.first()
            val otherGaps = gaps.drop(1)
            if (otherGaps.isEmpty()) return null   // need at least one more gap to compare
            val avgOtherGap = otherGaps.average()
            if (kotlin.math.abs(firstGap - 2 * avgOtherGap) > avgOtherGap * 0.5) return null
        }

        return pairs
    }

    /**
     * Returns the raw pair list without any validation. Useful for debugging.
     * (Still uses the original constants – only for debugging.)
     */
    fun getRawPairs(proj: FloatArray, offset: Int): List<Pair<Int, Int>> {
        val rawPeaks = PeakFinder.findRawPeaks(proj, offset)
        val merged = PeakFinder.mergeClosePeaks(rawPeaks)
        return pairEdges(merged, GridConstants.EXPECTED_WIDTH_MIN, GridConstants.EXPECTED_WIDTH_MAX)
    }

    /**
     * Greedy peak‑pairing within the given (minGap, maxGap) range.
     */
    fun pairEdges(
        peaks: List<Int>,
        minGap: Int,
        maxGap: Int,
        maxLookAhead: Int = 10
    ): List<Pair<Int, Int>> {
        val pairs = mutableListOf<Pair<Int, Int>>()
        val used = BooleanArray(peaks.size)
        for (i in peaks.indices) {
            if (used[i]) continue
            for (j in i + 1 until minOf(i + maxLookAhead, peaks.size)) {
                val gap = peaks[j] - peaks[i]
                if (gap in minGap..maxGap && !used[j]) {
                    pairs.add(peaks[i] to peaks[j])
                    used[i] = true
                    used[j] = true
                    break
                }
            }
        }
        return pairs
    }
}